File: EmailService.WebApi/Models/EmailModels.cs

using System.ComponentModel.DataAnnotations;


namespace EmailService.WebApi.Models
{
    /// <summary>
    /// Request to send email with TMS EmailHtml as body and another TMS export as attachment, each with its own property values
    /// </summary>
    public class SendEmailWithTmsHtmlAndAttachmentRequest
    {
        public string? FromAccount { get; set; }
        [Required]
        public List<string> ToRecipients { get; set; } = new();
        public List<string> CcRecipients { get; set; } = new();
        public List<string> BccRecipients { get; set; } = new();
        [Required]
        public string Subject { get; set; } = string.Empty;
        /// <summary>
        /// TMS template ID for the email body (EmailHtml)
        /// </summary>
        [Required]
        public Guid BodyTemplateId { get; set; }
        /// <summary>
        /// Property values for the email body (EmailHtml)
        /// </summary>
        [Required]
        public Dictionary<string, string> BodyPropertyValues { get; set; } = new();
        /// <summary>
        /// TMS template ID for the attachment (can be the same or different from body)
        /// </summary>
        [Required]
        public Guid AttachmentTemplateId { get; set; }
        /// <summary>
        /// Property values for the attachment
        /// </summary>
        [Required]
        public Dictionary<string, string> AttachmentPropertyValues { get; set; } = new();
        /// <summary>
        /// Export format for the attachment (not EmailHtml)
        /// </summary>
        [Required]
        public TmsExportFormat AttachmentExportFormat { get; set; }
    }
    public class EmailAccount
    {
        public string Name { get; set; } = string.Empty;
        public string DisplayName { get; set; } = string.Empty;
        public string EmailAddress { get; set; } = string.Empty;
        public bool IsDefault { get; set; } = false;
    }

    public class SmtpConfiguration
    {
        public string Host { get; set; } = string.Empty;
        public int Port { get; set; }
        public bool EnableSsl { get; set; }
        public string Username { get; set; } = string.Empty;
        public string Password { get; set; } = string.Empty;
    }

    public class EmailConfiguration
    {
        public SmtpConfiguration Smtp { get; set; } = new();
        public List<EmailAccount> Accounts { get; set; } = new();
    }

    public enum EmailStatus
    {
        Pending = 0,
        Sent = 1,
        Failed = 2
    }

    public enum TmsExportFormat
    {
        Original = 0,
        Word = 1,
        Html = 2,
        EmailHtml = 3,
        Pdf = 4
    }

    public class SendEmailWithTemplateRequest
    {
        /// <summary>
        /// The email account to send from (optional - uses default if not specified)
        /// </summary>
        public string? FromAccount { get; set; }
        
        /// <summary>
        /// List of primary recipients
        /// </summary>
        [Required]
        public List<string> ToRecipients { get; set; } = new();
        
        /// <summary>
        /// List of CC recipients (optional)
        /// </summary>
        public List<string> CcRecipients { get; set; } = new();
        
        /// <summary>
        /// List of BCC recipients (optional)
        /// </summary>
        public List<string> BccRecipients { get; set; } = new();
        
        /// <summary>
        /// Email subject
        /// </summary>
        [Required]
        public string Subject { get; set; } = string.Empty;
        
        /// <summary>
        /// Plain text email body (optional)
        /// </summary>
        public string? PlainTextBody { get; set; }
        
        /// <summary>
        /// HTML email body (optional, ignored if ExportFormat is EmailHtml)
        /// </summary>
        public string? HtmlBody { get; set; }
        
        /// <summary>
        /// TMS template ID to generate content from
        /// </summary>
        [Required]
        public Guid TemplateId { get; set; }
        
        /// <summary>
        /// Property values to inject into the template
        /// </summary>
        [Required]
        public Dictionary<string, string> PropertyValues { get; set; } = new();
        
        /// <summary>
        /// Export format for the generated document
        /// EmailHtml = replaces email body, others = attachment
        /// </summary>
        [Required]
        public TmsExportFormat ExportFormat { get; set; }
    }

    public class SendEmailWithDocumentsRequest
    {
        /// <summary>
        /// The email account to send from (optional - uses default if not specified)
        /// </summary>
        public string? FromAccount { get; set; }
        
        /// <summary>
        /// List of primary recipients
        /// </summary>
        [Required]
        public List<string> ToRecipients { get; set; } = new();
        
        /// <summary>
        /// List of CC recipients (optional)
        /// </summary>
        public List<string> CcRecipients { get; set; } = new();
        
        /// <summary>
        /// List of BCC recipients (optional)
        /// </summary>
        public List<string> BccRecipients { get; set; } = new();
        
        /// <summary>
        /// Email subject
        /// </summary>
        [Required]
        public string Subject { get; set; } = string.Empty;
        
        /// <summary>
        /// Plain text email body (optional)
        /// </summary>
        public string? PlainTextBody { get; set; }
        
        /// <summary>
        /// HTML email body (optional)
        /// </summary>
        public string? HtmlBody { get; set; }
        
        /// <summary>
        /// List of CMS document IDs to attach
        /// </summary>
        public List<Guid> CmsDocumentIds { get; set; } = new();
    }

    public class EmailSendResponse
    {
        /// <summary>
        /// Unique identifier for this email send operation
        /// </summary>
        public Guid EmailId { get; set; }
        
        /// <summary>
        /// Success or error message
        /// </summary>
        public string Message { get; set; } = string.Empty;
        
        /// <summary>
        /// Current status of the email
        /// </summary>
        public EmailStatus Status { get; set; }
        
        /// <summary>
        /// Timestamp when the email was processed
        /// </summary>
        public DateTime SentAt { get; set; }
        
        /// <summary>
        /// Detailed error message if sending failed
        /// </summary>
        public string? ErrorMessage { get; set; }
    }

    /// <summary>
    /// Request to test an email template with full support for all body types and attachments
    /// </summary>
    public class TestEmailTemplateRequest
    {
        [Required]
        public Guid TemplateId { get; set; }

        [Required]
        public List<string> ToRecipients { get; set; } = new();
        
        public List<string> CcRecipients { get; set; } = new();
        public List<string> BccRecipients { get; set; } = new();

        /// <summary>
        /// Property values for TMS body generation (if template uses TMS for body)
        /// </summary>
        public Dictionary<string, string>? TmsBodyPropertyValues { get; set; }

        /// <summary>
        /// Property values for each TMS attachment by attachment index
        /// Key: attachment index (0-based), Value: property values for that TMS template
        /// </summary>
        public Dictionary<int, Dictionary<string, string>>? TmsAttachmentPropertyValues { get; set; }
    }
}

File: Manteq-DOC-UI/projects/manteq-ui-lib/src/lib/models/email.models.ts
/**
 * Email body source type enum
 */
export enum EmailBodySourceType {
  PlainText = 0,
  TmsTemplate = 1,
  CustomTemplate = 2
}

/**
 * Attachment source type enum
 */
export enum AttachmentSourceType {
  CmsDocument = 1,
  TmsTemplate = 2,
  CustomFile = 3
}

/**
 * TMS export format enum
 */
export enum TmsExportFormat {
  Original = 0,
  Word = 1,
  Html = 2,
  EmailHtml = 3,
  Pdf = 4
}

/**
 * Email template entity
 */
export interface EmailTemplate {
  id: string;
  name: string;
  subject: string;
  htmlContent: string;
  plainTextContent?: string;
  templateId?: string;
  bodySourceType: EmailBodySourceType;
  tmsTemplateId?: string;
  customTemplateFilePath?: string;
  isActive: boolean;
  category?: string;
  sentCount: number;
  failureCount: number;
  createdBy: string;
  createdDate: string;
  totalAttempts: number;
  successRate: number;
  attachments?: EmailTemplateAttachment[];
}

/**
 * Email template attachment entity
 */
export interface EmailTemplateAttachment {
  id: string;
  emailTemplateId: string;
  sourceType: AttachmentSourceType;
  cmsDocumentId?: string;
  tmsTemplateId?: string;
  tmsExportFormat?: number;
  customFilePath?: string;
  customFileName?: string;
  displayOrder: number;
}

/**
 * Create email template request
 */
export interface CreateEmailTemplateRequest {
  name: string;
  subject: string;
  htmlContent: string;
  plainTextContent?: string;
  templateId?: string;
  category?: string;
  bodySourceType: EmailBodySourceType;
  tmsTemplateId?: string;
  customTemplateFilePath?: string;
}

/**
 * Update email template request
 */
export interface UpdateEmailTemplateRequest {
  name?: string;
  subject?: string;
  htmlContent?: string;
  plainTextContent?: string;
  templateId?: string;
  category?: string;
  bodySourceType?: EmailBodySourceType;
  tmsTemplateId?: string;
  customTemplateFilePath?: string;
}

/**
 * TMS attachment request
 */
export interface TmsAttachmentRequest {
  templateId: string;
  exportFormat: TmsExportFormat;
  propertyValues: Record<string, string>;
  customFilename?: string;
}

/**
 * Send email with flexible attachments request
 */
export interface SendEmailRequest {
  fromAccount?: string;
  toRecipients: string[];
  ccRecipients?: string[];
  bccRecipients?: string[];
  subject: string;
  emailTemplateId: string;
  bodyPropertyValues?: Record<string, string>;
  cmsDocumentIds?: string[];
  tmsAttachments?: TmsAttachmentRequest[];
}

/**
 * Send email with template request
 */
export interface SendEmailWithTemplateRequest {
  fromAccount?: string;
  toRecipients: string[];
  ccRecipients: string[];
  bccRecipients: string[];
  subject: string;
  plainTextBody?: string;
  htmlBody?: string;
  templateId: string;
  TmsBodyPropertyValues: { [key: string]: string };
  exportFormat: number;
}

/**
 * Send email with documents request
 */
export interface SendEmailWithDocumentsRequest {
  fromAccount?: string;
  toRecipients: string[];
  ccRecipients: string[];
  bccRecipients: string[];
  subject: string;
  plainTextBody?: string;
  htmlBody?: string;
  cmsDocumentIds: string[];
}

/**
 * Email send response
 */
export interface EmailSendResponse {
  emailId: string;
  message: string;
  status: EmailStatus;
  sentAt: Date | string;
  errorMessage?: string;
}

/**
 * Email status enum
 */
export enum EmailStatus {
  Pending = 0,
  Sent = 1,
  Failed = 2
}

/**
 * Email account
 */
export interface EmailAccount {
  name: string;
  displayName: string;
  email: string;
}

/**
 * Test email template request
 */
export interface TestEmailTemplateRequest {
  templateId: string;
  toRecipients: string[];
  ccRecipients: string[];
  bccRecipients: string[];
  tmsBodyPropertyValues?: Record<string, string>;
  tmsAttachmentPropertyValues?: Record<number, Record<string, string>>;
}

